### PR TITLE
Test ResourceTiming creation for some scenarios

### PR DESCRIPTION
**Pull request by noamr**
- Script RT entry should be created before load/error event
- Worker & worklet RT entry should be created
- <a ping> should create an RT entry

### CODE REVIEW

File: resource-timing/worklet-rt-entries.html [hunk 1]
```diff
@@ -0,0 +1,19 @@
...
+<title>Resource Timing Entry Sequence of Events for Worklets</title>
+<link rel="help" href="https://w3c.github.io/resource-timing/"/>
+<script src="/resources/testharness.js"></script>
+<script src="/resources/testharnessreport.js"></script>
+</head>
+<body>
+<script>
+    promise_test(async() => {
+        const url = new URL("/worklets/resources/empty-worklet-script.js", location.href).toString();
+        await CSS.paintWorklet.addModule(url);
```
Comment (tommckee1) [inline resource-timing/worklet-rt-entries.html:14]: Are there other worklet types we should be testing? I'm not sure which browsers support which worklets, tbh, but [MDN](https://developer.mozilla.org/en-US/docs/Web/API/Worklet) mentions
- PaintWorklet
- AudioWorklet
- AnimationWorklet
- LayoutWorklet

**Review by tommckee1 — CHANGES_REQUESTED** [review]
Thanks for tackling this Noam!

File: resource-timing/ping-rt-entries.html [hunk 2]
```diff
@@ -0,0 +1,33 @@
...
+        document.body.appendChild(link);
+        new PerformanceObserver(t.step_func(() => {
+            const entries = performance.getEntriesByName(ping);
+            assert_equals(entries.length, 1);
+            const [entry] = entries;
+            assert_equals(entry.initiatorType, 'ping');
+            assert_greater_than(entry.duration, delay);
+            t.done();
+        })).observe({"type": "resource"});
+        link.click();
```
[the remark this answers is not in the document]
Comment (noamr) [reply resource-timing/ping-rt-entries.html:29]: Ah yes let me refactor this piece

File: resource-timing/script-rt-entries.html [hunk 3]
```diff
@@ -0,0 +1,49 @@
...
+        }));
+        script.src = url;
+        document.body.appendChild(script);
+        t.add_cleanup(() => script.remove());
+    }, "The RT entry for a non-existent script should be available when the script 'error' event fires");
+
+    async_test(t => {
+        const url = new URL('resources/empty.js?worker', location.href).toString();
+        const myWorker = new Worker(url);
+        t.step_timeout(() => {
```
[the remark this answers is not in the document]
Comment (noamr) [reply resource-timing/script-rt-entries.html:39]: Actually I'll use the existing wait-for-entry thingy

File: resource-timing/worklet-rt-entries.html (hunk 1 above)
Comment (noamr) [reply resource-timing/worklet-rt-entries.html:14]: Yea I don't think they're really supported, I wanted to have a go with one type of the most common worklet to get started

File: resource-timing/ping-rt-entries.html [hunk 4]
```diff
@@ -0,0 +1,33 @@
...
+        link.setAttribute('href', 'resources/close.html');
+        link.setAttribute('target', '_blank');
+        link.setAttribute('ping', ping);
+        link.innerText = 'Link';
+        document.body.appendChild(link);
+        new PerformanceObserver(t.step_func(() => {
+            const entries = performance.getEntriesByName(ping);
+            assert_equals(entries.length, 1);
+            const [entry] = entries;
+            assert_equals(entry.initiatorType, 'ping');
```
Comment (tommckee1) [inline resource-timing/ping-rt-entries.html:25]: Is it spec'd somewhere that this needs to be "ping"? Looking at the [hyperlink auditing](https://html.spec.whatwg.org/multipage/links.html#hyperlink-auditing) section of the HTML spec doesn't mention an initiator type afaict... should it?

File: resource-timing/initiator-type/misc-script.html [hunk 5]
```diff
@@ -0,0 +1,29 @@
...
+<link rel="help" href="https://www.w3.org/TR/resource-timing-2/#dom-performanceresourcetiming-initiatortype"/>
+<script src="/resources/testharness.js"></script>
+<script src="/resources/testharnessreport.js"></script>
+<script src="/resource-timing/resources/observe-entry.js"></script>
+<script src="resources/initiator-type-test.js"></script>
+
+</head>
+
+<script>
+  const moduleWorkerURL = 'resources/empty.js?id=moduleWorker';
```
Comment (tommckee1) [inline resource-timing/initiator-type/misc-script.html:15]: nit: If we drop the "id=" part we'll be more consistent with the other URLs in this test.

File: resource-timing/initiator-type/misc-script.html [hunk 6]
```diff
@@ -0,0 +1,29 @@
...
+  const modulePreloadURL = 'resources/empty.js?modulePreload';
+  const workerURL = 'resources/empty.js?worker';
+  const link = document.createElement('link');
+  link.rel = 'modulepreload';
+  link.href = modulePreloadURL;
+  document.head.appendChild(link);
```
Comment (tommckee1) [inline resource-timing/initiator-type/misc-script.html:21]: Could we move this test case over to `resource-timing/intitiator-type/link.html`? We hope to keep `initiator-type` tests organized by their element.

File: resource-timing/initiator-type/misc-script.html [hunk 7]
```diff
@@ -0,0 +1,29 @@
+<!DOCTYPE html>
```
Comment (tommckee1) [inline resource-timing/initiator-type/misc-script.html:1]: If we remove the `link` test, I think we're only left with worker tests in this file so maybe we could call this `workers.html`?

**Comment by noamr** [comment]
@tommckee1: revised based on comment.

File: resource-timing/initiator-type/link.html [hunk 8]
```diff
@@ -18,7 +18,12 @@
     href="/resource-timing/resources/resource_timing_test0.css?id=preload">
 <link rel="prerender" href="/resource-timing/resources/green.html?id=prerender">
 <link rel="manifest" href="/resource-timing/resources/manifest.json">
+<link rel="modulepreload" href="resources/empty.js?id=modulePreload">
 <script>
+  const link = document.createElement('link');
+  link.rel = 'modulepreload';
+  link.href = modulePreloadURL;
+  document.head.appendChild(link);
```
Comment (tommckee1) [inline resource-timing/initiator-type/link.html:26]: I don't think we want this anymore, do we?
```suggestion
```

File: resource-timing/initiator-type/workers.html [hunk 9]
```diff
@@ -0,0 +1,23 @@
+<!DOCTYPE html>
+<html>
+<head>
+<meta charset="utf-8" />
+<title>Resource Timing initiator type: extra script types</title>
```
Comment (tommckee1) [inline resource-timing/initiator-type/workers.html:5]: Maybe
```suggestion
<title>Resource Timing initiatorType: worker resources</title>
```
?

File: resource-timing/worklet-rt-entries.html (hunk 1 above)
Comment (tommckee1) [reply resource-timing/worklet-rt-entries.html:14]: Sticking to PaintWorklet for now SGTM!

**Review by tommckee1 — APPROVED** [review]

**Comment by noamr** [comment]
@yoavweiss can you approve as well? Needs a reviewer with write access (I don't remember how to/who can grant it)

File: resource-timing/initiator-type/link.html [hunk 10]
```diff
@@ -30,6 +31,7 @@
   initiator_type_test("resource_timing_test0.css?id=preload", "link", "<link preload>");
   initiator_type_test("green.html?id=prerender", "link", "<link prerender>");
   initiator_type_test("manifest.json", "link", "<link manifest>");
+  initiator_type_test("resources/empty.js?id=modulePreload", "other", "module preload");
```
Comment (yoavweiss) [inline resource-timing/initiator-type/link.html:34]: Why "other"?

File: resource-timing/initiator-type/workers.html [hunk 11]
```diff
@@ -0,0 +1,23 @@
...
+<script src="/resource-timing/resources/observe-entry.js"></script>
+<script src="resources/initiator-type-test.js"></script>
+
+</head>
+
+<script>
+  const moduleWorkerURL = 'resources/empty.js?moduleWorker';
+  const workerURL = 'resources/empty.js?worker';
+  new Worker(moduleWorkerURL, {type: "module"});
+  new Worker(workerURL, {type: "module"});
```
Comment (yoavweiss) [inline resource-timing/initiator-type/workers.html:18]: Should this be "module"? Or "classic"? (or omitted)

**Review by yoavweiss — COMMENTED** [review]
Looks good! A couple of questions..

File: resource-timing/initiator-type/link.html (hunk 10 above)
Comment (noamr) [reply resource-timing/initiator-type/link.html:34]: That's a very good question. With initiator types, for now I'm spec'ing them based on what implementations currently do. I think once we have all the specs in place that reflect the status quo we should do an initiator type overhaul. Right now they don't serve their purpose as they show "other" exactly in the places where you'd want them to be meaningful.

File: resource-timing/initiator-type/workers.html (hunk 11 above)
Comment (noamr) [reply resource-timing/initiator-type/workers.html:18]: Oops, fixed.

File: resource-timing/initiator-type/link.html (hunk 10 above)
Comment (yoavweiss) [reply resource-timing/initiator-type/link.html:34]: That's fair. Can you add a comment indicating that one would expect this to be "link", but this is what implementations are currently doing?

**Review by yoavweiss — APPROVED** [review]
LGTM

**Pull request merged**